src/transformers/models/modernbert/modular_modernbert.py

```diff
@@ -0,0 +1,1249 @@
+        _init_modernbert_weights(self.config, self.Wqkv, module_type=ModernBertModuleType.in_module)
+        _init_modernbert_weights(self.config, self.Wo, module_type=ModernBertModuleType.out_module)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        position_ids: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        cu_seqlens: Optional[torch.Tensor] = None,
+        max_seqlen: Optional[int] = None,
+        **kwargs,
+    ) -> torch.Tensor:
+        """Perform self-attention.
+
+        There are two attention implementations supported: PyTorch's SDPA attention and Flash Attention 2.
+
+        The arguments are unpadded. The SDPA implementation of attention requires padded arguments while the
+        Flash Attention implementation does not. If using SDPA we first call `pad_input`. Once we compute
+        attention, we re-unpad our outputs for the other layers. The pad/unpad operations add overhead, but not
+        sending pad tokens through ffs saves compute.
+
+        Args:
+            hidden_states: (total_nnz, dim)
+            cu_seqlens: (batch + 1,)
+            max_seqlen: int
+            indices: (total_nnz,)
+            attn_mask: (batch, max_seqlen)
+
+        Returns:
+            attention: (total_nnz, dim)
+        """
+        bs, dim = hidden_states.shape[:2]
+        qkv = self.Wqkv(hidden_states)
+        qkv = qkv.view(-1, 3, self.num_heads, self.head_dim)
+
+        attn = MODERNBERT_ATTENTION_FUNCTION[self.config._attn_implementation](
+            self,
+            qkv=qkv,
+            rotary_emb=self.rotary_emb,
+            position_ids=position_ids,
+            attention_mask=attention_mask,
+            local_attention=self.local_attention,
+            cu_seqlens=cu_seqlens,
+            max_seqlen=max_seqlen,
+            bs=bs,
+            dim=dim,
+        )
+
+        return self.out_drop(self.Wo(attn))
+
+
+class ModernBertFlashAttention2(ModernBertAttention):
+    def __init__(self, config: ModernBertConfig, layer_id: Optional[int] = None):
+        super().__init__(config, layer_id)
+        self.config._attn_implementation = "flash_attention_2"
+        logger.warning_once(
+            "The `ModernBertFlashAttention2` class is deprecated in favor of simply modifying the `config._attn_implementation`"
+            "attribute of the `GemmaAttention` class! It will be removed in v4.48"
+        )
+
+
+class ModernBertSdpaAttention(ModernBertAttention):
+    def __init__(self, config: ModernBertConfig, layer_id: Optional[int] = None):
+        super().__init__(config, layer_id)
+        self.config._attn_implementation = "sdpa"
+        logger.warning_once(
+            "The `ModernBertFlashAttention2` class is deprecated in favor of simply modifying the `config._attn_implementation`"
+            "attribute of the `GemmaAttention` class! It will be removed in v4.48"
+        )
```

Comment:
I don't think these are necessary, they might just be leftover copy-pastes from Gemma.